### PR TITLE
Removal of IDigestionParams and ICommonParameters and addition of secondary PSM constructor

### PR DESCRIPTION
**Pull request by rmmiller22**
The interfaces of IDigestionParams and ICommonParameters have been removed and all references to these have been transitioned to DigestionParameters and CommonParameters class references.

In preparation for match between runs to be effective with multi-parameter search, the PeptideSpectralMatch class has a second constructor in which digestionParameters is an argument. This new constructor has been implemented in all search tasks utilizing PSMs.

### CODE REVIEW

File: GUI/GPTMDTaskWindow.xaml.cs [hunk 1]
```diff
@@ -275,15 +275,7 @@ private void SaveButton_Click(object sender, RoutedEventArgs e)
...
-                    MinPeptideLength = int.Parse(txtMinPeptideLength.Text, NumberStyles.Any, CultureInfo.InvariantCulture),
-                    MaxPeptideLength = int.Parse(txtMaxPeptideLength.Text, NumberStyles.Any, CultureInfo.InvariantCulture),
-                    Protease = (Protease)proteaseComboBox.SelectedItem,
-                    MaxModificationIsoforms = int.Parse(maxModificationIsoformsTextBox.Text, CultureInfo.InvariantCulture),
-                    InitiatorMethionineBehavior = (InitiatorMethionineBehavior)initiatorMethionineBehaviorComboBox.SelectedIndex
-                }
+                DigestionParams = new DigestionParams((Protease)proteaseComboBox.SelectedItem, int.Parse(missedCleavagesTextBox.Text, CultureInfo.InvariantCulture), int.Parse(txtMinPeptideLength.Text, NumberStyles.Any, CultureInfo.InvariantCulture), int.Parse(txtMaxPeptideLength.Text, NumberStyles.Any, CultureInfo.InvariantCulture), int.Parse(maxModificationIsoformsTextBox.Text, CultureInfo.InvariantCulture), (InitiatorMethionineBehavior)initiatorMethionineBehaviorComboBox.SelectedIndex)
```
Comment (rmillikin) [inline GUI/GPTMDTaskWindow.xaml.cs]: can you put a line break after each comma here to make it a bit more readable

File: TaskLayer/MetaMorpheusTask.cs [hunk 2]
```diff
@@ -41,10 +41,10 @@ public abstract class MetaMorpheusTask
                             .WithConversionFor<TomlString>(convert => convert
                                 .ToToml(custom => custom.ToString())
                                 .FromToml(tmlString => GlobalVariables.ProteaseDictionary[tmlString.Value])))
-                        .ConfigureType<ICommonParameters>(ct => ct
+                        .ConfigureType<CommonParameters>(ct => ct
                             .CreateInstance(() => new CommonParameters()))
-                        .ConfigureType<IDigestionParams>(ct => ct
-                            .CreateInstance(() => new DigestionParams()))
+                        .ConfigureType<DigestionParams>(ct => ct
+                            .CreateInstance(() => new DigestionParams(new Protease("weird",null,null,TerminusType.C,CleavageSpecificity.Full,null,null,null))))
```
Comment (rmillikin) [inline TaskLayer/MetaMorpheusTask.cs]: do you need to specify all of these in the constructor?

File: TaskLayer/MetaMorpheusTask.cs [hunk 3]
```diff
@@ -343,21 +341,25 @@ protected static void WritePsmsToTsv(IEnumerable<PeptideSpectralMatch> items, st
                 {
                     output.WriteLine(heh.ToString(ModstoWritePruned));
                 }
-            }
+           }
         }
-
-        protected static HashSet<IDigestionParams> GetListOfDistinctDigestionParams(ICommonParameters commonParameters, IEnumerable<ICommonParameters> enumerable)
+        protected static HashSet<DigestionParams> GetListOfDistinctDigestionParams(CommonParameters commonParameters, IEnumerable<CommonParameters> enumerable)
```
Comment (rmillikin) [inline TaskLayer/MetaMorpheusTask.cs]: does this method need to exist? isn't it the same as `HashSet<DigestionParams>(commonParams.Select(p => p.DigestionParams))`

File: TaskLayer/NeoSearchTask/NeoSearchTask.cs [hunk 4]
```diff
@@ -25,13 +25,8 @@ public NeoSearchTask() : base(MyTask.Neo)
...
             NeoParameters = new NeoParameters();
 
-            IDigestionParams tempDigParams = new DigestionParams
-            {
-                MinPeptideLength = 8,
-                MaxPeptideLength = 13,
-                Protease = GlobalVariables.ProteaseDictionary["non-specific"],
-                MaxMissedCleavages = 12
-            };
+            DigestionParams tempDigParams = new DigestionParams(GlobalVariables.ProteaseDictionary["non-specific"], 12, 8, 13, 1024, InitiatorMethionineBehavior.Variable, 2, false, TerminusType.N);
```
Comment (rmillikin) [inline TaskLayer/NeoSearchTask/NeoSearchTask.cs]: this is too many parameters specified I think, try to make very minimal changes to the previous structure

File: TaskLayer/SearchTask/SearchTask.cs [hunk 5]
```diff
@@ -407,9 +408,9 @@ private static void WriteIndexEngineParams(IndexingEngine indexEngine, string fi
         private static string GenerateOutputFolderForIndices(List<DbForTask> dbFilenameList)
         {
             var folder = Path.Combine(Path.GetDirectoryName(dbFilenameList.First().FilePath), DateTime.Now.ToString("yyyy-MM-dd-HH-mm-ss", CultureInfo.InvariantCulture));
-            Directory.CreateDirectory(folder);
+                Directory.CreateDirectory(folder);
```
Comment (rmillikin) [inline TaskLayer/SearchTask/SearchTask.cs]: remove indent

File: TaskLayer/SearchTask/SearchTask.cs [hunk 6]
```diff
@@ -173,14 +176,15 @@ protected override MyTaskResults RunSpecific(string OutputFolder, List<DbForTask
                             GenerateIndexes(indexEngine, dbFilenameList, ref peptideIndex, ref fragmentIndex, taskId);
                         }
 
+
```
Comment (rmillikin) [inline TaskLayer/SearchTask/SearchTask.cs]: remove these extra lines everywhere

File: Test/AddCompIonsTest.cs [hunk 7]
```diff
@@ -40,14 +40,10 @@ public static void TestAddCompIonsClassic()
...
+            Protease protease = new Protease("Custom Protease", new List<string> { "K" }, new List<string>(), TerminusType.C, CleavageSpecificity.Full, null, null, null);
             CommonParameters CommonParameters = new CommonParameters
             {
-                DigestionParams = new DigestionParams
-                {
-                    Protease = new Protease("Custom Protease", new List<string> { "K" }, new List<string>(), TerminusType.C, CleavageSpecificity.Full, null, null, null),
-                    MinPeptideLength = 1,
-                    MaxMissedCleavages = 0
-                },
+                DigestionParams = new DigestionParams(protease, 0, 1),
```
Comment (rmillikin) [inline Test/AddCompIonsTest.cs]: revise to be same format as old

File: Test/AddCompIonsTest.cs [hunk 8]
```diff
@@ -100,18 +96,16 @@ public static void TestCompIons_ModernSearch()
...
+            Protease protease = new Protease("singleN", new List<string> { "K" }, new List<string>(), TerminusType.C, CleavageSpecificity.Full, null, null, null);
             CommonParameters CommonParameters = new CommonParameters
             {
                 PrecursorMassTolerance = new PpmTolerance(5),
-                DigestionParams = new DigestionParams
-                {
-                    Protease = new Protease("singleN", new List<string> { "K" }, new List<string>(), TerminusType.C, CleavageSpecificity.Full, null, null, null),
-                    MinPeptideLength = 1,
-                },
+                DigestionParams = new DigestionParams(protease, 2,1),
```
Comment (rmillikin) [inline Test/AddCompIonsTest.cs]: revise to be same format as old

File: Test/AmbiguityTest.cs [hunk 9]
```diff
@@ -17,13 +17,10 @@ internal static class AmbiguityTest
...
         {
+            Protease protease = new Protease("Custom Protease", new List<string> { "K" }, new List<string>(), TerminusType.C, CleavageSpecificity.Full, null, null, null);
             CommonParameters CommonParameters = new CommonParameters
             {
-                DigestionParams = new DigestionParams
-                {
-                    Protease = new Protease("Custom Protease", new List<string> { "K" }, new List<string>(), TerminusType.C, CleavageSpecificity.Full, null, null, null),
-                    MinPeptideLength = 1,
-                },
+                DigestionParams = new DigestionParams(protease,2,1),
```
Comment (rmillikin) [inline Test/AmbiguityTest.cs]: revise to be same format as old

File: Test/AnalysisEngineTest.cs [hunk 10]
```diff
@@ -21,15 +21,10 @@ public static class AnalysisEngineTests
...
             CommonParameters CommonParameters = new CommonParameters
             {
-                DigestionParams = new DigestionParams
-                {
-                    Protease = new Protease("Custom Protease", new List<string> { "K" }, new List<string>(), TerminusType.C, CleavageSpecificity.Full, null, null, null),
-                    MinPeptideLength = 1,
-                    MaxMissedCleavages = 0,
-                    MaxModificationIsoforms = 1042,
-                },
+                DigestionParams = new DigestionParams(protease,0,1,int.MaxValue,1042),
```
Comment (rmillikin) [inline Test/AnalysisEngineTest.cs]: revise to be same format as old

File: Test/FdrTest.cs [hunk 11]
```diff
@@ -28,7 +28,7 @@ public static void FdrTestMethod()
             List<string> nestedIds = new List<string>();
 
             Protein p = new Protein("MNKNNKNNNKNNNNK", null);
-            DigestionParams digestionParams = new DigestionParams();
+            DigestionParams digestionParams = new DigestionParams(GlobalVariables.ProteaseDictionary["trypsin"]);
```
Comment (rmillikin) [inline Test/FdrTest.cs]: is this change needed?

File: Test/IndexEngineTest.cs [hunk 12]
```diff
@@ -37,17 +37,15 @@ public static void TestIndexEngine()
...
 
+            Protease p = new Protease("Custom Protease", new List<string> { "K" }, new List<string>(), TerminusType.C, CleavageSpecificity.Full, null, null, null);
             CommonParameters CommonParameters = new CommonParameters
             {
-                DigestionParams = new DigestionParams
-                {
-                    Protease = new Protease("Custom Protease", new List<string> { "K" }, new List<string>(), TerminusType.C, CleavageSpecificity.Full, null, null, null),
-                    MinPeptideLength = 1,
-                },
+                DigestionParams = new DigestionParams(p,2,1),
```
Comment (rmillikin) [inline Test/IndexEngineTest.cs]: you get the idea

File: TaskLayer/MetaMorpheusTask.cs (hunk 3 above)
Comment (rmmiller22) [reply TaskLayer/MetaMorpheusTask.cs]: CommonParameters does not allow .Select method

File: TaskLayer/GPTMDTask/GPTMDTask.cs [hunk 13]
```diff
@@ -103,7 +103,8 @@ protected override MyTaskResults RunSpecific(string OutputFolder, List<DbForTask
             {
                 newDatabases = new List<DbForTask>()
             };
-            HashSet<IDigestionParams> ListOfDigestionParams = GetListOfDistinctDigestionParams(CommonParameters, fileSettingsList.Select(b => SetAllFileSpecificCommonParams(CommonParameters, b)));
+            var fileSpecificCommonParams = fileSettingsList.Select(b => SetAllFileSpecificCommonParams(CommonParameters, b));
+            HashSet<DigestionParams> ListOfDigestionParams = new HashSet<DigestionParams>(fileSpecificCommonParams.Select(p => p.DigestionParams));
```
Comment (rmillikin) [inline TaskLayer/GPTMDTask/GPTMDTask.cs]: 👍

File: GUI/CalibrateTaskWindow.xaml.cs [hunk 14]
```diff
@@ -228,12 +228,13 @@ private void SaveButton_Click(object sender, RoutedEventArgs e)
...
-            digestionParamsToSave.MaxPeptideLength = int.Parse(txtMaxPeptideLength.Text, NumberStyles.Any, CultureInfo.InvariantCulture);
-            digestionParamsToSave.Protease = (Protease)proteaseComboBox.SelectedItem;
-            digestionParamsToSave.MaxModificationIsoforms = int.Parse(maxModificationIsoformsTextBox.Text, CultureInfo.InvariantCulture);
+           
+            Protease protease = (Protease)proteaseComboBox.SelectedItem;
+            int MaxMissedCleavages = int.Parse(missedCleavagesTextBox.Text, CultureInfo.InvariantCulture);
+            int MinPeptideLength = int.Parse(txtMinPeptideLength.Text, NumberStyles.Any, CultureInfo.InvariantCulture);
+            int MaxPeptideLength= int.Parse(txtMaxPeptideLength.Text, NumberStyles.Any, CultureInfo.InvariantCulture);
+            int MaxModificationIsoforms= int.Parse(maxModificationIsoformsTextBox.Text, CultureInfo.InvariantCulture);
+            DigestionParams digestionParamsToSave = new DigestionParams(protease.Name, MaxMissedCleavages, MinPeptideLength, MaxPeptideLength, MaxModificationIsoforms);
```
Comment (rmillikin) [inline GUI/CalibrateTaskWindow.xaml.cs]: make sure you're including the specifiers in all of these:
`new DigestionParams(Protease: protease.Name, ........`
